packages/chevrotain/src/parse/parser/traits/recognizer_engine.ts

```diff
@@ -210,35 +213,30 @@ export class RecognizerEngine {
     this.shortRuleNameToFull[shortName] = ruleName
     this.fullRuleNameToShort[ruleName] = shortName
 
-    function invokeRuleWithTry(this: MixedInParser, args: any[]) {
+    function invokeRuleWithTry(this: MixedInParser, ...args: ARGS[]): R {
       try {
+        this.ruleInvocationStateUpdate(shortName, ruleName, this.subruleIdx)
+        this.subruleIdx = 0
         if (this.outputCst === true) {
           impl.apply(this, args)
           const cst = this.CST_STACK[this.CST_STACK.length - 1]
           this.cstPostRule(cst)
-          return cst
+          return cst as unknown as R
         } else {
           return impl.apply(this, args)
         }
       } catch (e) {
-        return this.invokeRuleCatch(e, resyncEnabled, recoveryValueFunc)
+        return this.invokeRuleCatch(e, resyncEnabled, recoveryValueFunc) as R
       } finally {
         this.ruleFinallyStateUpdate()
       }
     }
 
-    const wrappedGrammarRule = function (
-      this: MixedInParser,
-      idxInCallingRule: number = 0,
-      args: any[]
-    ) {
-      this.ruleInvocationStateUpdate(shortName, ruleName, idxInCallingRule)
-      return invokeRuleWithTry.call(this, args)
-    }
+    const wrappedGrammarRule: ParserMethod<ARGS, R> = Object.assign(
```

Comment:
there is no longer a wrapper here, we can remove this const 
and add the properties directly on `invokeRuleWithTry`

Reply:
When you do so, please ensure to separate the **definition** of `invokeWithTry` and adding the extra properties.
There is no need to pack too many operations into a single expression / statement as it can reduce readability.

Reply:
Going to leave it for now as we have const foo = ..;  return foo everywhere here. It shouldn't matter from a perf perspective as the minifier should inline it.

Reply:
it is not from a performance concern as the creation of the grammar only happens once per rule.
Only readability concern...

But you can leave it as is because now that you simplified the `idx` handling I may be able to implement other optimizations